fdbserver/storageserver.actor.cpp

```diff
@@ -1611,6 +1612,67 @@ void merge(Arena& arena,
 	}
 }
 
+ACTOR Future<Optional<Value>> quickGetValue(StorageServer* data, StringRef key, Version version) {
+	if (data->shards[key]->isReadable()) {
+		try {
+			// TODO: Use a lower level API may be better? Or tweak priorities?
+			GetValueRequest req(Span().context, key, version, Optional<TagSet>(), Optional<UID>());
+			data->actors.add(data->readGuard(req, getValueQ));
+			GetValueReply reply = wait(req.reply.getFuture());
+			return reply.value;
+		} catch (Error& e) {
+			std::cout << "quickGetValue fallback because of exception " << e.name() << std::endl;
+		}
+	} else {
+		std::cout << "quickGetValue fallback because not in the server" << std::endl;
+	}
+
+	state Transaction tr(data->cx);
+	tr.setVersion(version);
+	// TODO: is DefaultPromiseEndpoint the best priority for this?
+	tr.info.taskID = TaskPriority::DefaultPromiseEndpoint;
+	Future<Optional<Value>> valueFuture = tr.get(key, Snapshot::True);
+	// TODO: async in case it needs to read from other servers.
```

Comment:
Can we reply error if it needs to hop to another server and let client side retry? 

Doing like this means the first SS is doing the same amount of work client does if the second-hop key is not colocated with the first-hop key.

Reply:
Yes, that's another design choice. See the discussion in the design doc.

I'm currently using this one because it has the simplest API and least modification in client.

Reply:
@xumengpanda I would prefer first SS "do the work".  If we can bypass JNI and not have to serde index records, it is a win

Reply:
ah, I forgot the JNI overhead when I wrote the comment. 


Letting the first SS forward the requests and do the work can avoid JNI and is likely better. 
I'd like to point out its tradeoff as well:
* Suppose the data flow is client->SS1 (to find index) -> SS2 (to find the value) -> SS2 directly reply to client, which means the first SS forwards the request to SS2
* SS1 will need to cache the shard-to-SS mapping, which causes extra memory.  That means we use some extra memory on storage node to save some cpu time on compute node.  

We probably don't have to decide the corner case now.

Reply:
There are three possibilities, I believe:

1. first SS does all the proxy calls to wherever
2. C client does any proxy calls when hopping within SS is not possible
3. API client does any proxy calls

3 has the JNI overhead, but 2 doesn't. It's more modification, as @nblintao says, but I think it might be the most robust. Particularly once we undertake to do range scans of the hopped-to place.